#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QMouseEvent>

MainWindow::MainWindow(QWidget *parent) : QWidget(parent),
                                          ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    tmpCmd = QString();
    ui->biaoqingFrame->setVisible(false);
    ui->cover_label->setVisible(false);
    //å»çª—å£è¾¹æ¡†

    setWindowFlags(Qt::FramelessWindowHint | windowFlags());

}
//æµ‹è¯•ç”¨å‡½æ•°ï¼Œæ­£å¼å‘å¸ƒåŠ¡å¿…åˆ é™¤ï¼ˆç›¸å…³æ§ä»¶
void MainWindow::on_pushButton_2_clicked()
{
    if (ui->listWidget->count() == 0)
        return;
    delete ui->listWidget->item(ui->listWidget->count() - 1);
}
//æµ‹è¯•ç”¨å‡½æ•°ï¼Œæ­£å¼å‘å¸ƒåŠ¡å¿…åˆ é™¤ï¼ˆç›¸å…³æ§ä»¶
void MainWindow::on_pushButton_clicked()
{
    QString msg = ui->textEdit->toPlainText();
    pushMessageIntoChatWindow(false, msg, QString::number(QDateTime::currentDateTime().toTime_t()), false);
    qDebug() << QString::number(QDateTime::currentDateTime().toTime_t());
}

void MainWindow::clearAllMessage()
{
    while (ui->listWidget->count() > 0)
    {
        delete ui->listWidget->item(ui->listWidget->count() - 1);
    }
}

MainWindow::MainWindow(QString ud, QTcpSocket *sock, QWidget *parent) : QWidget(parent),
                                                                        ui(new Ui::MainWindow)
{
    udata = ud;
    client = sock;
    ui->setupUi(this);
    ui->biaoqingFrame->setVisible(false);
    ui->cover_label->setVisible(false);
    userList.clear();
    imag = new WNCimage(ud, client);
    //å»çª—å£è¾¹æ¡†
    setWindowFlags(Qt::FramelessWindowHint | windowFlags());
    // TODOï¼šæ²¡æœ‰çª—å£è¾¹æ¡†æ€ä¹ˆé¡¶éƒ¨æ‹–åŠ¨ï¼Ÿç»§æ‰¿QLabelï¼Œé‡å†™mouseMoveEventå’Œpress...ï¼Œé¡¶éƒ¨ä¸¤ä¸ªlabelå°±å¯ä»¥å®ç°æ‹–åŠ¨çª—å£äº†
    connect(client, SIGNAL(readyRead()), this, SLOT(hadreadyread()));
    qDebug() << "ud=" << ud;
    //å‘é€å¥½å‹åˆ—è¡¨è¯·æ±‚
    std::string data = Encoder_askfriendsList(udata.toStdString());
    QString packData = QString::fromStdString(data);
    client->write((packData.toUtf8()));
}

MainWindow::~MainWindow()
{
    delete ui;
}

int MainWindow::packData(const QString &pack, Json::Value &data)
{
    Json::Reader reader;
    Json::Value jtmp;
    tmpCmd.append(pack);
    if (!reader.parse(tmpCmd.toStdString(), jtmp))
    {
        //        qDebug() << "æœåŠ¡å™¨è¿”å›ä¿¡æ¯é”™è¯¯\n";
        return 0;
    }
    if (!jtmp.isObject())
    {
        //        qDebug() << "æœåŠ¡å™¨è¿”å›ä¿¡æ¯é”™è¯¯\n";
        return 0;
    }
    if (jtmp.isObject())
    {
        data = jtmp;
        tmpCmd.clear();
        return 1;
    }
}

void MainWindow::hadreadyread()
{

    QByteArray recvArray = client->read(1048576);
    //åŒºåˆ†æ˜¯chatä¿¡æ¯è¿˜æ˜¯å¥½å‹åˆ—è¡¨ä¿¡æ¯
    QString backdata = recvArray;
    qDebug() << backdata;

    Json::Value jtmp, tmp_info;

    if (!packData(backdata, jtmp))
    {
        return;
    }

    Json::FastWriter writer;
    //    qDebug()<<QString::fromStdString(writer.write(jtmp));
    qDebug() << "type = " << QString::fromStdString(jtmp["type"].asString());
    if (jtmp["type"].asString() == "askfriendsList")
    { //å¤„ç†å¥½å‹åˆ—è¡¨ä¿¡æ¯
        std::string username, info;
        bool state;
        std::string userimage;
        if (Decoder_askfriendsList(jtmp["info"], username, userList, userimage) == 0)
        {
            qDebug("askfriends data back from server error/n");
            return;
        }
        /*ä½¿ç”¨userListç”Ÿæˆå¥½å‹åˆ—è¡¨*/
        user_image = QString2Qimage(QString::fromStdString(userimage));
        /*Todo:æŠŠå¸¦å¥½å‹å¤´åƒä¿¡æ¯çš„userListå¯¼è¿›å»ï¼Œæ˜¾ç¤ºå¤´åƒ
         *
         *
         *
         */
        // friendList = new Ui::FriendList(ui->friendList, userList);
        initConnection();
    }
    else if (jtmp["type"].asString() == "chat")
    { //å¤„ç†èŠå¤©ä¿¡æ¯
        std::string sender_username, msg;
        if (Decoder_chat(jtmp["info"], sender_username, msg) == 0)
        {
            qDebug("chat data back from server error/n");
            return;
        }
        pushMessageIntoChatWindow(false, QString::fromStdString(msg), QString::number(QDateTime::currentDateTime().toTime_t()), false);
    }
    else if (jtmp["type"].asString() == "addfriends")
    { // mainwindowé‡Œå¤„ç†åŠ å¥½å‹çš„è¿”å›ä¿¡æ¯
        std::string friend_username;
        bool state;
        std::string friendimage;
        if (Decoder_addfriends(jtmp["info"], friend_username, state, friendimage) == 0)
        {
            qDebug("addfriends data back from server error/n");
            return;
        }
        if (state == 1)
        {
            /*Todo:åœ¨åˆ—è¡¨é‡Œæ·»åŠ è¿™ä¸ªæ–°çš„å¥½å‹
             *
             *
             */
            QMessageBox::information(this, "æç¤º", ("æˆåŠŸæ·»åŠ å¥½å‹" + friend_username).c_str()); //
            add->close();
        }
        else
            QMessageBox::information(this, "æç¤º", "è¯¥å¥½å‹ä¸å­˜åœ¨");
    }
    else if (jtmp["type"].asString() == "ready_chatfile")
    {
        bool state;
        if (Decoder_ready_chatfile(jtmp["info"], state) == 0)
        {
            qDebug("ready_chatfile data back from server error/n");
            return;
        }
        if (state == 1)
        {
            /*Todo :æ ¹æ®å¥½å‹åˆ—è¡¨ç”Ÿæˆè¦å‘é€ä¿¡æ¯å¥½å‹usersList
             * è¿™é‡Œæš‚æ—¶ç”¨hxyä»£æ›¿
             *
             */
            std::vector<std::string> userslist; // userListéœ€è¦ä»å¥½å‹æ å¯¼å…¥çš„ï¼Œè¿™é‡Œå›ºå®šçš„å ä¸ªä½ç½®
            userslist.push_back("hxy");
            /* Todoï¼šå¯¼å…¥è¦å‘é€ç»™çš„userList
             *
             *
             */
            imag->send_chatfile_image(userslist);
        }
        else
        {
            QMessageBox::information(this, "æç¤º", "å‘é€æ–‡ä»¶å¤±è´¥");
        }
    }
    else if (jtmp["type"].asString() == "ready_submit_image")
    {
        bool state;
        if (Decoder_ready_submit_image(jtmp["info"], state) == 0)
        {
            qDebug("ready_submit_image data back from server error/n");
            return;
        }
        if (state == 1)
        {
            imag->send_submit_image();
        }
        else
        {
            QMessageBox::information(this, "æç¤º", "å‘é€æ–‡ä»¶å¤±è´¥");
        }
    }
    else if (jtmp["type"].asString() == "chatfile")
    {

        std::string sender_username, info, fn;
        if (Decoder_chatfile(jtmp["info"], fn, sender_username, info) == 0)
        {
            qDebug("chatfile data back from server error/n");
            return;
        }
        QString filename = QString::fromStdString(fn);
        QString cat = filename.right(filename.size() - filename.lastIndexOf(".") - 1);
        qDebug() << "æ–‡ä»¶ç±»å‹ï¼š" << cat;
        if (cat == "jpg" || cat == "jpeg" || cat == "svg" || cat == "bmp" || cat == "png" || cat == "gif")
        {

            QImage chat_image = QString2Qimage(QString::fromStdString(info));
            //å¤„ç†è¿”å›æ¥çš„chat_imageå›¾ç‰‡ä¿¡æ¯
            pushImageIntoChatWindow(false, chat_image, QString::number(QDateTime::currentDateTime().toTime_t()));
            return;
        }
        // Todo:ä¹‹åå¯ä»¥åŠ å…¶ä»–ç±»å‹æ–‡ä»¶çš„å¤„ç†
    }
    else if (jtmp["type"].asString() == "submit_image")
    {
        std::string username;
        bool state;
        if (Decoder_submit_image(jtmp["info"], username, state) == 0)
        {
            qDebug("submit_image data back from server error/n");
            return;
        }
        /* Todo: æ‰“å°user_imageåˆ°å„ç§åœ°æ–¹
         *
         *
         *
         *
         */
    }
    else
    {
        qDebug() << "æœåŠ¡å™¨è¿”å›ä¿¡æ¯é”™è¯¯\n";
        return;
    }
}

void MainWindow::on_hideButton_clicked()
{
    showMinimized();
}

void MainWindow::on_closeButton_clicked()
{
    close();
}

void MainWindow::on_send_clicked()
{

    QString msg = ui->textEdit->toPlainText();
    if (msg.isEmpty())
        return; //ç©ºå­—ç¬¦ä¸å‘é€
    /*Todo :æ ¹æ®å¥½å‹åˆ—è¡¨ç”Ÿæˆè¦å‘é€ä¿¡æ¯å¥½å‹usersList
     * è¿™é‡Œæš‚æ—¶ç”¨hxyä»£æ›¿
     *
     */

    std::vector<std::string> usersList;
    usersList.push_back(chattingInfo.chatFriend.userName);
    //å‘é€æ•°æ®åè®®
    std::string data = Encoder_chat(udata.toStdString(), msg.toStdString(), usersList);
    QString packData = QString::fromStdString(data);
    client->write((packData.toLocal8Bit()));
    //æµ‹è¯•å‘æ•°æ®
    pushMessageIntoChatWindow(true, msg, QString::number(QDateTime::currentDateTime().toTime_t()), false);
    ui->textEdit->clear();
}

/*
 * pushMessageIntoChatWindowå‡½æ•°è¯´æ˜
 * bool typeï¼šå‘ç»™åˆ«äººæ¶ˆæ¯ï¼Œå³ä¾§æ°”æ³¡==trueï¼›æ¥æ”¶åˆ°æ¶ˆæ¯ï¼Œå·¦ä¾§æ°”æ³¡==falseã€‚
 * QString msgï¼šå‘é€/æ¥æ”¶çš„æ–‡æœ¬ã€‚
 * QString timeï¼šæ¶ˆæ¯æ—¶é—´ã€‚
 *      å‚è€ƒï¼štime = QString::number(QDateTime::currentDateTime().toTime_t());
 * bool isSending = falseï¼šæ˜¯å¦æ­£åœ¨å‘é€ï¼Œå¤§æ¦‚ç‡ç”¨ä¸åˆ°æ‰€ä»¥å…ˆè®¾ç½®æˆfalse
 *      ä»…å‘é€æ¶ˆæ¯æœ‰æ•ˆï¼Œæ¥æ”¶æ¶ˆæ¯æ—¶è¿™é‡Œå¯ä»¥éšä¾¿èµ‹å€¼ã€‚
 * QImage* image = NULL:å¤´åƒæŒ‡é’ˆï¼Œé»˜è®¤ä¸ºNULLï¼Œåˆ™æ˜¾ç¤ºé»˜è®¤å¤´åƒã€‚
 *
 * *******æ³¨æ„********
 * â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“ç‰¹åˆ«æ³¨æ„â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“â†“
 * â‘    UIä¿®æ”¹æ—¶éœ€è¦å…¨å±€æœç´¢æ›¿æ¢widthWithoutListWidgetçš„å€¼
 *      è¯¥å€¼æ˜¯æ•´ä¸ªmainWindowå»æ‰ListWidgetå¯¹è¯æ¡†çš„å®½åº¦
 *      ä¸æ›¿æ¢å¯èƒ½å¯¹è¯æ¡†æ˜¾ç¤ºå¤§å°å¼‚å¸¸
 * â‘¡   TODOï¼šæ£€æŸ¥å¯èƒ½æœ‰å†…å­˜æº¢å‡ºï¼Œæ³¨æ„delete ui->listWidget->item(i);
 *      forå¾ªç¯å°†æ‰€æœ‰éƒ½åˆ æ‰
 * â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘â†‘
 * æ³¨æ„å¤„ç†ç©ºæ¶ˆæ¯ï¼Œç©ºæ¶ˆæ¯æœ€å¥½åˆ«ä¼ è¿›æ¥ï¼Œä¼ è¿›æ¥ä¹Ÿä¸ä¼šæ˜¾ç¤º
 * å¦‚æœæ˜¯å‘æ¶ˆæ¯ï¼Œæ³¨æ„æ¸…ç©ºè¾“å…¥æ¡†
 * TODOï¼šä¸æ¸…æ¥šåˆ†è¾¨ç‡å’Œç¼©æ”¾çš„å½±å“ï¼Œå¯èƒ½ä¼šæœ‰å°æ¦‚ç‡é•¿æ–‡æœ¬æ˜¾ç¤ºæ¨ªå‘æ‹–åŠ¨æ¡
 */
void MainWindow::pushMessageIntoChatWindow(bool type, QString msg, QString time, QImage *image, bool isSending)
{

    if (msg == "")
    {
        // TODOï¼šæ˜¯ä¸æ˜¯è¦æŠ¥ä¸ªé”™ï¼Ÿ
        return;
    }
    qDebug() << "addMessage" << msg << time << ui->listWidget->count();
    if (type)
    {
        if (isSending)
        {
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealMessage(messageW, item, msg, time, ChatMessageWidget::User_Me, image);
        }
        else
        {
            bool isOver = true;
            for (int i = ui->listWidget->count() - 1; i > 0; i--)
            {
                ChatMessageWidget *messageW = (ChatMessageWidget *)ui->listWidget->itemWidget(ui->listWidget->item(i));
                if (messageW->text() == msg)
                {
                    isOver = false;
                    messageW->setTextSuccess();
                }
            }
            if (isOver)
            {
                dealMessageTime(time);
                ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
                QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
                dealMessage(messageW, item, msg, time, ChatMessageWidget::User_Me, image);
                messageW->setTextSuccess();
            }
        }
    }
    else
    {
        dealMessageTime(time);
        ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
        dealMessage(messageW, item, msg, time, ChatMessageWidget::User_She, image);
    }
    ui->listWidget->setCurrentRow(ui->listWidget->count() - 1);
}

void MainWindow::dealMessage(ChatMessageWidget *messageW, QListWidgetItem *item, QString text, QString time, ChatMessageWidget::User_Type type, QImage *image)
{
    messageW->setFixedWidth(ui->listWidget->width() - 25);
    QSize size = messageW->fontRect(text);
    item->setSizeHint(size);
    messageW->setText(text, time, size, type, image);
    ui->listWidget->setItemWidget(item, messageW);
}

void MainWindow::dealMessageTime(QString curMsgTime)
{
    bool isShowTime = false;
    if (ui->listWidget->count() > 0)
    {
        QListWidgetItem *lastItem = ui->listWidget->item(ui->listWidget->count() - 1);
        ChatMessageWidget *messageW = (ChatMessageWidget *)ui->listWidget->itemWidget(lastItem);
        int lastTime = messageW->time().toInt();
        int curTime = curMsgTime.toInt();
        qDebug() << "curTime lastTime:" << curTime - lastTime;
        isShowTime = ((curTime - lastTime) > 60); // ä¸¤ä¸ªæ¶ˆæ¯ç›¸å·®ä¸€åˆ†é’Ÿ
                                                  //        isShowTime = true;
    }
    else
    {
        isShowTime = true;
    }
    if (isShowTime)
    {
        ChatMessageWidget *messageTime = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *itemTime = new QListWidgetItem(ui->listWidget);

        QSize size = QSize(ui->listWidget->width() - 25, 40);
        messageTime->resize(size);
        itemTime->setSizeHint(size);
        messageTime->setText(curMsgTime, curMsgTime, size, ChatMessageWidget::User_Time);
        ui->listWidget->setItemWidget(itemTime, messageTime);
    }
}

void MainWindow::on_pushButton_addfriend_clicked()
{
    add = new searchFriends(udata, client);
    add->show();
}
void MainWindow::on_pushButton_image_clicked()
{
    QString image_addr = QFileDialog::getOpenFileName(this);
    if (image_addr.length() == 0)
    {
        return;
    }
    imag->getImagefromdir(image_addr);
    imag->sendinform_submit_image();
    user_image = imag->image;
    changeMyIcon();
}
void MainWindow::changeMyIcon(QImage * uimg){
    if(uimg == NULL){
        uimg = new QImage(":/assets/defaultHead.png");
    }
    QIcon qicon(QPixmap::fromImage(*uimg));
    ui->pushButton_image->setIcon(qicon);
}

void MainWindow::on_pushButton_send_image_clicked()
{
    QString image_addr = QFileDialog::getOpenFileName(this);
    if (image_addr.length() == 0)
    {
        return;
    }
    imag->getImagefromdir(image_addr);

    imag->sendinform_chatfile_image();
    pushImageIntoChatWindow(true, imag->image, QString::number(QDateTime::currentDateTime().toTime_t()));
}
void MainWindow::initConnection()
{
    connect(friendList, SIGNAL(openChatroom(QVariant)), this, SLOT(startChatting(QVariant)));
}

void MainWindow::startChatting(QVariant variant)
{
    clearAllMessage();

    Ui::User chatFriend = variant.value<Ui::User>();
    qDebug() << "here";
    qDebug() << QString(chatFriend.userName.c_str());
    ui->chatName->setText("   " + QString(chatFriend.userName.c_str()));
    chattingInfo.chatFriend = chatFriend;

    std::vector<std::string> messages;
    messages.push_back("user1/ä½ å¥½/1661088000");
    messages.push_back("111/ä½ å¥½/1661088100");
    messages.push_back("user1/å¾ˆé«˜å…´è®¤è¯†ä½ /1661088200");
    messages.push_back("111/æˆ‘ä¹Ÿæ˜¯/1661088290");
    chattingInfo.record = new MessageRecord(messages);
    for (std::string rec : chattingInfo.record->getAllMessageRecord())
    {
        QStringList list = QString(rec.c_str()).split("/");
        if (list[0] == udata)
        {
            //ä¼ è‡ªå·±å¤´åƒçš„ Qimage
            pushMessageIntoChatWindow(true, list[1], list[2]);
        }
        else
        {
            //ä¼ å¯¹æ–¹å¤´åƒçš„QImage
            pushMessageIntoChatWindow(false, list[1], list[2]);
        }
    }
}

//ä»¥ä¸‹æ˜¯æ‹–åŠ¨çª—å£çš„ä»£ç 
void MainWindow::mousePressEvent(QMouseEvent *ev)
{
    if (ev->button() == Qt::LeftButton)
    {
        //æ±‚åæ ‡å·®å€¼
        //å½“å‰ç‚¹å‡»åæ ‡-çª—å£å·¦ä¸Šè§’åæ ‡
        p = ev->globalPos() - this->frameGeometry().topLeft();
    }
}

void MainWindow::mouseMoveEvent(QMouseEvent *ev)
{
    if (ev->buttons() & Qt::LeftButton && p.x() != 0 && p.y() != 0)
    {
        //ç§»åˆ°å·¦ä¸Šè§’
        move(ev->globalPos() - p);
    }
}

void MainWindow::mouseReleaseEvent(QMouseEvent *ev)
{
    p.setX(0);
    p.setY(0);
}

/*
 * pushMessageIntoChatWindowå‡½æ•°è¯´æ˜
 * bool typeï¼šå‘ç»™åˆ«äººæ¶ˆæ¯ï¼Œå³ä¾§æ°”æ³¡==trueï¼›æ¥æ”¶åˆ°æ¶ˆæ¯ï¼Œå·¦ä¾§æ°”æ³¡==falseã€‚
 * QImage msgï¼šå‘é€/æ¥æ”¶çš„Qimageã€‚
 * QString timeï¼šæ¶ˆæ¯æ—¶é—´ã€‚
 *      å‚è€ƒï¼štime = QString::number(QDateTime::currentDateTime().toTime_t());
 * bool isSending = falseï¼šæ˜¯å¦æ­£åœ¨å‘é€ï¼Œå¤§æ¦‚ç‡ç”¨ä¸åˆ°æ‰€ä»¥å…ˆè®¾ç½®æˆfalse
 *      ä»…å‘é€æ¶ˆæ¯æœ‰æ•ˆï¼Œæ¥æ”¶æ¶ˆæ¯æ—¶è¿™é‡Œå¯ä»¥éšä¾¿èµ‹å€¼ã€‚
 * QImage* image = NULL:å¤´åƒæŒ‡é’ˆï¼Œé»˜è®¤ä¸ºNULLï¼Œåˆ™æ˜¾ç¤ºé»˜è®¤å¤´åƒã€‚
 *
 */
void MainWindow::pushImageIntoChatWindow(bool type, QImage msg, QString time, QImage *image, bool isSending)
{
    if (type)
    {
        if (isSending)
        {
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealImage(messageW, item, msg, time, ChatMessageWidget::Image_Me, image);
        }
        else
        {
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealImage(messageW, item, msg, time, ChatMessageWidget::Image_Me, image);
            messageW->setTextSuccess();
        }
    }
    else
    {
        dealMessageTime(time);
        ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
        dealImage(messageW, item, msg, time, ChatMessageWidget::Image_She, image);
    }
    ui->listWidget->setCurrentRow(ui->listWidget->count() - 1);
}

void MainWindow::dealImage(ChatMessageWidget *messageW, QListWidgetItem *item, QImage img, QString time, ChatMessageWidget::User_Type type, QImage *image)
{
    messageW->setFixedWidth(ui->listWidget->width() - 25);
    int w = img.width(), h = img.height();
    QSize size = messageW->fontRect(QString("---------------\n---------------\n---------------\n---------------\n---------------\n---------------\n-"));
    qDebug() << size.width() << "...." << size.height();
    //    QSize size(300,150);
    item->setSizeHint(size);
    messageW->setImage(img, time, type, image);
    ui->listWidget->setItemWidget(item, messageW);
}

void MainWindow::on_biaoqingButton_clicked()
{
    ui->biaoqingFrame->setVisible(true);

    QImage qi = ui->bButton1->icon().pixmap(ui->bButton1->icon().actualSize(QSize(32, 32))).toImage();

    pushFileIntoChatWindow(true, "j:\\å¾®ä¿¡å›¾ç‰‡_20220211210810.jpg","å¾®ä¿¡å›¾ç‰‡ï¼ˆæ–‡ä»¶åï¼‰",QString::number(QDateTime::currentDateTime().toTime_t()));




}


void MainWindow::pushFileIntoChatWindow(bool type,QString filePath, QString msg, QString time, QImage *image, bool isSending)
{

    if (msg == "")
    {
        // TODOï¼šæ˜¯ä¸æ˜¯è¦æŠ¥ä¸ªé”™ï¼Ÿ
        return;
    }
    qDebug() << "addFile" << msg << time << ui->listWidget->count();
    if (type)
    {
        if (isSending)
        {
            dealMessageTime(time);
            ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
            QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
            dealFile(messageW, item,filePath ,msg, time, ChatMessageWidget::File_Me, image);
        }
        else
        {
            bool isOver = true;
            for (int i = ui->listWidget->count() - 1; i > 0; i--)
            {
                ChatMessageWidget *messageW = (ChatMessageWidget *)ui->listWidget->itemWidget(ui->listWidget->item(i));
                if (messageW->text() == msg)
                {
                    isOver = false;
                    messageW->setTextSuccess();
                }
            }
            if (isOver)
            {
                dealMessageTime(time);
                ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
                QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
                dealFile(messageW, item,filePath, msg, time, ChatMessageWidget::File_Me, image);
                messageW->setTextSuccess();
            }
        }
    }
    else
    {
        dealMessageTime(time);
        ChatMessageWidget *messageW = new ChatMessageWidget(ui->listWidget->parentWidget());
        QListWidgetItem *item = new QListWidgetItem(ui->listWidget);
        dealFile(messageW, item,filePath, msg, time, ChatMessageWidget::User_She, image);
    }
    ui->listWidget->setCurrentRow(ui->listWidget->count() - 1);
}

void MainWindow::dealFile(ChatMessageWidget *messageW, QListWidgetItem *item,QString filePath, QString text, QString time, ChatMessageWidget::User_Type type, QImage *image)
{
    text = QString("ã€ğŸ“‚åŒå‡»æ‰“å¼€æ–‡ä»¶ã€‘")+text;
    messageW->setFixedWidth(ui->listWidget->width() - 25);
    QSize size = messageW->fontRect(text);
    item->setSizeHint(size);
    messageW->setFile(filePath,text, time, size, type, image);
    ui->listWidget->setItemWidget(item, messageW);
}
